src/4.3_to_5.0/migrateSettingsFolder.ts

```diff
@@ -76,6 +78,44 @@ function migrateSettingsMap(legacySettingsMap: {
     );
   }
 
+  /* Smart filtering settings start. */
+  const isSmartFilteringDisabled =
+    legacySettingsMap["memberSelection.smartFiltering"] === false;
+  if (isSmartFilteringDisabled) {
+    migratedSettingsMap["smartFiltering.isEnabled"] = false;
+  } else {
+    const nonEmptyEvaluationMeasureUniqueName =
+      legacySettingsMap["memberSelection.smartFiltering.nonEmptyMeasure"];
+    if (nonEmptyEvaluationMeasureUniqueName !== undefined) {
+      const [, nonEmptyEvaluationMeasureName] = unquote(
+        nonEmptyEvaluationMeasureUniqueName,
+      );
+      // @ts-expect-error TypeScript rightfully complains that this setting key does not exist in 5.0
+      // It was added only in 5.1 indeed, but it makes sense to migrate it as an intermediate step.
```

Comment:
Why does it make sense? Wouldn't the fact that the migration script yields settings that do nothing be seen as a bug by customers?

Why not do it in `5.0_to_5.1/migrateSettingsFolder`?

Reply:
The `nonEmptyEvaluationMeasureName` setting exists in 4.3 and 5.1, but not in 5.0.
You do want to preserve it if you go from 4.3 to 5.1.

Implementation-wise, right now the migration script works only with iterative steps (4.3 to 5.0, 5.0 to 5.1), etc.
So if you discard the 4.3 setting while going from 4.3 to 5.0, you would not end up having it in the 5.1 output.

**Alternative 1**
Indeed do not discard it when going from 4.3 to 5.0, but actually migrate it to its new syntax only in the 5.0 to 5.1 step. But there would be a type error as well, as the 4.3 key does not exist in the 5.0 settings.

**Alternative 2**
Do an extra step outside of the iterative steps to avoid the type error.

But:
- We would start doing things like `if fromVersion === and toVersion === then....` in `migrateContentServer` which was supposed to be generic. This is a bit what I'm doing for calculated measures though.
- We would explode the logic of migrate settings in several places.

No solution is ideal theoretically, but pragmatically the current state of the PR seemed like the best option to me.

Reply:
I agree that Alternative 2 is bad.

But I picture a customer asking why the setting that we generated during their migation to 5.0 does not work in 5.0 and I wouldn't be able to tell them _"yes it's on purpose, it's because of how our migration script is made internally and I did not want to add a ts-expect-error"_. In practice, I would have to agree that it is a bug and we would have to fix it (=> to backport the 5.1 setting to 5.0).

I still remember the feedback from SCB at the London connect with their 2 main issues:
- upgrades are two costly
- the product is so frequently bugged that they feel like we don't do QA at all and that we consider our customers as our QA testers

I generally don't want to introduce something that I consider right off the bat, and this kind of face-to-face feedback probably makes it even more important.

So I like Alternative 1 even if it means some `ts-expect-error`.

Or I can think of an Alternative 3, where the step 4.3 -> 5.0 transforms the setting key into something that makes it clear that it does not work in 5.0 and will work in 5.1. For instance `__upgrade_to_5.1_to_use_this__<setting_key>`. And then make the 5.0 -> 5.1 step migrate it again to what it needs.

Finally Alternative 4 which might be the best: actually backport the setting right now so that this problem does not exist.

What do you think?